Mi 486 horizontal scroll

nensidosari opened the pull request:
## Changes

<!--
### Describe what this PR does

- Short and concise, bullet points can help
- Screenshots if applicable can also help
-->

## Events

Did you introduce any new tracking events?

<!--
If yes please remove the comment HTML comment tags and fill the table below

Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type   | event_name  | value |
|--------|-------------|-------|
| Change/New | event name  | extra: { ... } |

-->

## Experiment

Did you introduce any new experiments?

<!--
If yes please remove the comment HTML comment tags and follow the instructions below

Don't forget to send a message to the [#experiments](https://dailydotdev.slack.com/archives/C02JAUF8HJL/p1715175315620999) channel, following the template in slack, and adding a link to the message here.

> [!IMPORTANT]
> Please do not merge the PR until the experiment enrolment is approved.

-->

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

<!--
If relevant, please remove the comment HTML comment tags and fill the checkboxes below

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

-->

<!--
Copy and paste the below line outside the HTML comment tags to link this PR to the ticket in Jira

AS-{number} #done
or
MI-{number} #done
-->

rebelchris commented on packages/shared/src/components/HorizontalScroll/useHorizontalScrollHeader.tsx:
```diff
@@ -0,0 +1,127 @@
+  const [numCards, setNumCards] = useState<number>(0);
+
+  // Calculate the width of elements and the number of visible cards
+  const calculateVisibleCards = useCallback(() => {
```
This seems like a highly expensive calculation right?
do we really need to do it as a "seperate one" meaning, maybe we can just use the computedStyle of the main container? (as you already know how many children you render)

rebelchris commented on packages/shared/src/components/HorizontalScroll/useScrollManagement.tsx:
```diff
@@ -0,0 +1,40 @@
+
+  const [debouncedOnScroll] = useDebounceFn(checkScrollPosition, 100);
+
+  useEffect(() => {
```
Probably yes, but this needs to be a effect?
(Eg we cant leverage useCallback as we technically only add event listener?)

capJavert replied:
We can probably use useEventListener and I would use requestAnimationFrame instead (or in addition) to debounce to make it smooth and not impact scrolling.

nensidosari replied:
Addressed feedback as @capJavert suggested

nensidosari replied:
Based on Ante's comment

> hey all good, one thing I might have missed is that we don't really need requestAnimationFrame since you are not doing any layout manipulation directly, you are just reading values, so if not much of a hassle I think remove it to not complicate the code, sorry if going back and forth 

removing requestAnimationFrame additions

capJavert replied:
Approved as well 🙌

rebelchris reviewed (approved):
Looks to work perfectly!
So code wise happy, just wondering if we can optimize the heavy calculation.

Also edge case (which we had already, but maybe can fix now?)
If there's exact amount of items (4/4) the right button is still active:
https://mi-486-horizontal-scroll.preview.app.daily.dev/sources/TechWithTim (the second row)

nensidosari commented:
Also edge case (which we had already, but maybe can fix now?)
If there's exact amount of items (4/4) the right button is still active:
https://mi-486-horizontal-scroll.preview.app.daily.dev/sources/TechWithTim (the second row)

I know 😬 I tried to fix today morning, it works ok on storybook, but on the feed it messes up a little bit because how the scrollWidth is calculated in first load. So need to think a better solution for it but didn't want to block the whole thing.

ilasw commented on packages/shared/src/components/HorizontalScroll/HorizontalScroll.tsx:
```diff
@@ -0,0 +1,47 @@
+          className,
+        )}
+        role="region"
+        aria-label="Horizontal Scroll"
```
If we have always the `<Header />` component with the title for this section maybe we would like to use `aria-labelledby` pointing to the title. 
So user that enter into this section with a screen reader will hear something like: `Related posts` instead `Horizontal Scroll`. What do you think?

Reference: https://www.aditus.io/aria/aria-label/#aria-label-vs-aria-labelledby

nensidosari replied:
Thank u!! Will change :D

sshanzel commented on packages/shared/src/components/HorizontalScroll/HorizontalScroll.tsx:
```diff
@@ -0,0 +1,49 @@
+  const titleId = `horizontal-scroll-title-${Math.random()}`;
+
+  return (
+    <>
```
I think we should make this an actual wrapper of the header and content to make it more predictable.
```suggestion
    <div className="flex flex-col">
```

nensidosari replied:
Yeah makes sense, thank u! will add

sshanzel reviewed (approved):
Don't have anything to add on top of what is already mentioned, everything looked amazing. Love the fact that we have proper tests.

Just a reminder to move the hooks to their proper folders before merging after addressing the concerns.

ilasw commented on packages/shared/src/components/HorizontalScroll/HorizontalScroll.tsx:
```diff
@@ -0,0 +1,49 @@
+    onClickSeeAll,
+  });
+
+  const titleId = `horizontal-scroll-title-${Math.random()}`;
```
Thanks for updating! Just a small suggestion!

```suggestion
  const id = useId();
  const titleId = `horizontal-scroll-title-${id}`;
```

nensidosari commented:
As for this: Just a reminder to move the hooks to their proper folders before merging after addressing the concerns.

@sshanzel 

I'm a bit in doubt of wether that is the best case or not, and don't want to go against our setup. But feels like these hooks are not hooks that can be used on any occasion, they are tightly coupled to HorizontalScroll in general. So would it make sense to leave here?

Not sure what others think as well, @dailydotdev/web-team

capJavert commented:
@nensidosari I agree with hooks staying where they are, they are internal hooks for the component that we use only there, I did the same thing for SearchPanel hooks.
